File: packages/nodejs/src/domains/auth/oidc.ts
import jwt, { Algorithm, JwtPayload, TokenExpiredError } from 'jsonwebtoken';
import { Issuer, generators, Client, CallbackParamsType } from 'openid-client';
import { getDebug } from '../../logging';
import { signJwt } from './jwt';
const debug = getDebug('domains:auth:oidc');
import {
  forbidden,
  invalidOidcIdToken,
  signinFailed,
  unsupportedScope,
  faildDecodeOidcIdToken,
  mismatchKidOidcIdToken,
} from '../../errors';
import {
  AdminUserCreatePayload,
  findOneWithCredentialByEmail,
} from '../adminuser';
import {
  AdminUserSsoToken,
  AdminUserSsoTokenCreatePayload,
  list as findSsoTokens,
  upsertOne as upsertAdminUserSsoToken,
  updateOneByClientIdAndUserId as updateAdminUserSsoTokenOneByUserId,
} from '../adminuserssotoken';
import {
  createFirstAdminUser,
  formatTokenSetToSsoTokens,
  createViewer,
} from './common';
import {
  AUTH_PROVIDER,
  AUTH_TYPE,
  OIDC_DEFAULT_SCOPES,
  OIDC_TOKEN_REFRESH_BEFORE_SEC as OIDC_TOKEN_REFRESH_BUFFER_SEC,
} from '../../constants';
import http from 'http';

export interface OidcClientConfig {
  clientId: string;
  clientSecret: string;
  issuerUrl: string;
}

export interface OidcConfig extends OidcClientConfig {
  additionalScopes?: string[];
  userHostedDomains?: string[];
}

// キャッシュされたOIDCクライアント
let cachedOidcClient: Client | null = null;

// OIDCクライアントの生成
export const genOidcClient = async (
  config: OidcConfig,
  redirectUri?: string,
  forceReload?: boolean
): Promise<Client> => {
  // forceReloadがtrueでない場合にキャッシュを使用
  if (cachedOidcClient && !forceReload) {
    debug('Use cached oidcClient');
    return cachedOidcClient;
  }

  // OIDCプロバイダーのIssuerを取得
  const issuer = await Issuer.discover(
    config.issuerUrl + '/.well-known/openid-configuration'
  );
  debug('Discovered issuer %o', issuer);

  // issuer.metadata.scopes_supportedでサポートされていないスコープがないかチェック
  // https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery
  // RECOMMENDED. JSON array containing a list of the OAuth 2.0 [RFC6749] scope values that this server supports. The server MUST support the openid scope value. Servers MAY choose not to advertise some supported scope values even when this parameter is used, although those defined in [OpenID.Core] SHOULD be listed, if supported.
  // 推奨プロパティなのでない場合もある
  const scopesSupported = issuer.metadata.scopes_supported
    ? (issuer.metadata.scopes_supported as string[])
    : [];
  // scopes_supportedがある場合はチェック
  if (scopesSupported.length > 0) {
    // 追加スコープがある場合は追加スコープを含めてチェック
    const scopes =
      config.additionalScopes && config.additionalScopes.length > 0
        ? OIDC_DEFAULT_SCOPES.concat(config.additionalScopes)
        : OIDC_DEFAULT_SCOPES;
    // scopes の中のどれか一つでもサポートされていない場合はエラー
    if (scopes.some((scope) => !scopesSupported.includes(scope))) {
      throw unsupportedScope();
    }
  }

  debug('redirectUri %s', redirectUri);

  // クライアントの作成とキャッシュ
  const oidcClient = new issuer.Client({
    client_id: config.clientId,
    client_secret: config.clientSecret,
    response_types: ['code'],
    ...(redirectUri && { redirect_uris: [redirectUri] }),
  });

  // redirectUriが指定されている場合のみキャッシュ
  if (redirectUri) {
    cachedOidcClient = oidcClient;
  }

  return oidcClient;
};

// OIDC用のコードベリファイアを生成
export const genOidcCodeVerifier = async (): Promise<string> => {
  return generators.codeVerifier();
};

// Oidc認可画面URLを取得
export const getOidcAuthorizationUrl = async (
  config: OidcConfig,
  client: Client,
  codeVerifier: string,
  state: string
): Promise<string> => {
  // PKCE用のコードベリファイアを生成
  const codeChallenge = generators.codeChallenge(codeVerifier);

  // 認証URLを生成
  const authorizationUrl = client.authorizationUrl({
    scope:
      config.additionalScopes && config.additionalScopes.length > 0
        ? OIDC_DEFAULT_SCOPES.concat(config.additionalScopes).join(' ')
        : OIDC_DEFAULT_SCOPES.join(' '),
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
    state,
  });

  debug('Authorization URL:', authorizationUrl);

  return authorizationUrl;
};

// Oidcサインイン
export const signinOidc = async (
  req: http.IncomingMessage,
  client: Client,
  codeVerifier: string,
  redirectUri: string,
  params: CallbackParamsType,
  config: OidcConfig,
  multipleAuthUser: boolean
): Promise<string> => {
  debug('params:', params);
  debug('codeVerifier:', codeVerifier);

  const tokenSet = await client.callback(redirectUri, params, {
    code_verifier: codeVerifier,
    state: params.state,
  });

  const claims = tokenSet.claims();
  const ssoToken = formatTokenSetToSsoTokens(tokenSet);

  if (ssoToken.idToken === '') {
    debug(
      'signinOidc invalid authentication codeVerifier. %s, %o',
      codeVerifier,
      ssoToken
    );
    throw invalidOidcIdToken();
  }

  // emailチェック
  const email = claims.email;
  if (!email) {
    debug(
      'signinOidc invalid login claims: %o, idToken: %s',
      claims,
      tokenSet.id_token
    );
    throw invalidOidcIdToken();
  }
  // emailドメインチェック
  const emailDomain = email.split('@').pop() as string;
  if (
    config.userHostedDomains?.length &&
    !config.userHostedDomains.includes(emailDomain)
  ) {
    // 許可されていないメールドメイン
    debug('signinOidc illegal user email: %s', email);
    throw forbidden();
  }

  // トークンの有効期限が切れている場合は403
  if (tokenSet.expired()) {
    debug('Token expired!');
    throw forbidden();
  }

  // emailでユーザーを検索
  let adminUser = await findOneWithCredentialByEmail(email);

  // SSOトークンのUpsert
  const ssoTokenPayload = {
    authType: AUTH_TYPE.OIDC,
    provider: AUTH_PROVIDER.CUSTOM,
    clientId: config.clientId,
    ...ssoToken,
  } as AdminUserSsoTokenCreatePayload;

  // ユーザーが存在しない場合は新規作成
  if (!adminUser) {
    const adminUserCreatePayload: AdminUserCreatePayload = { email };

    // 最初ログイン時ユーザー作成(SUPER)
    adminUser = await createFirstAdminUser(
      AUTH_TYPE.OIDC,
      adminUserCreatePayload,
      ssoTokenPayload
    );

    // 管理者ユーザー存在する場合はviewerユーザー作成
    if (!adminUser) {
      adminUser = await createViewer(
        AUTH_TYPE.OIDC,
        adminUserCreatePayload,
        ssoTokenPayload
      );
    }
  } else {
    // userが存在する場合
    // multipleAuthUser=falseの場合は認証方法は1つのみに制限する
    if (!multipleAuthUser) {
      // 登録済みユーザーの認証方法の確認
      // password認証が登録済みの場合はエラー
      if (adminUser.password && adminUser.password !== '') {
        console.error('signinFailed password auth already', adminUser.id);
        throw signinFailed();
      }
      // userIdで登録済みSSOトークン情報を取得
      const ssoTokens = await findSsoTokens({ userId: adminUser.id });
      for (const st of ssoTokens.list) {
        // 今回のclientIDと違うSSOトークンがある場合はエラー
        if (st.clientId !== claims.aud) {
          console.error('signinFailed clientId already');
          throw signinFailed();
        }
      }
    }

    // SSOトークンのUserIDを設定
    ssoTokenPayload.userId = adminUser.id;

    // ここまででエラーがない場合はSSOトークンのUpsert
    upsertAdminUserSsoToken(ssoTokenPayload);
  }

  debug('signinOidc Sign jwt for user: %s', adminUser.id);
  return signJwt(adminUser.id, req);
};

// アクセストークンの検証
export const verifyOidcAccessToken = async (
  client: Client,
  config: OidcConfig,
  clientId: string,
  userId: string,
  ssoToken: AdminUserSsoToken
): Promise<boolean> => {
  // IDトークンの検証
  // vironではIDトークンはDBに保存されているので、改竄されることはないが
  // DBはviron利用者が管理するため改竄されることを考慮してvironLibとしてはIDトークンの検証を毎度行う
  const verified = await verifyOidcIdToken(
    client,
    config,
    ssoToken.idToken
  ).catch((e: Error) => {
    debug('verifyOidcIdToken failure. userId: %s, err: %o', userId, e);
    return e;
  });
  // IDトークンの検証に失敗した場合はエラー
  if (verified instanceof Error) {
    return false;
  }

  // リフレッシュトークンがない場合はIDトークンの有効期限だけで判定
  if (!ssoToken.refreshToken) {
    debug('verifyOidcAccessToken no refresh token. userId: %s', userId);
    // IDトークンの有効期限が切れている場合はエラー
    return !verified.expired;
  }

  // ---- 以下リフレッシュトークンがある場合の処理 ----
  // IDトークンの有効期限が近いもしくは期限切れ場合以外はリフレッシュしない
  if (!isRefresh(verified.payload.exp)) {
    debug('verifyOidcAccessToken no need to refresh token. userId: %s', userId);
    return true;
  }

  // リフレッシュトークンがある場合はリフレッシュトークンを使ってトークンを更新
  const refreshToken = ssoToken.refreshToken;
  const tokenset = await client.refresh(refreshToken).catch((e: Error) => {
    debug('AccessToken refresh failure. userId: %s, err: %o', userId, e);
    return e;
  });
  // アクセストークンのリフレッシュに失敗した場合はエラー
  if (tokenset instanceof Error) {
    return false;
  }
  debug('AccessToken refresh success! userId: %s', userId);

  // トークンを更新
  const newSsoToken = {
    ...formatTokenSetToSsoTokens(tokenset),
    userId,
    clientId,
    authType: AUTH_TYPE.OIDC,
    provider: AUTH_PROVIDER.CUSTOM,
  } as AdminUserSsoTokenCreatePayload;

  await updateAdminUserSsoTokenOneByUserId(clientId, userId, newSsoToken);
  return true;
};

/**
 * IDトークンの検証
 * 有効期限切れ以外の例外はスローする
 *
 * @param client
 * @param config
 * @param idToken
 * @returns {Promise<{decodedJwt: Jwt, isExpired: boolean}>} デコードしたJWTと期限切れ判定を返す
 */
export const verifyOidcIdToken = async (
  client: Client,
  config: OidcConfig,
  idToken: string
): Promise<{ payload: JwtPayload; expired: boolean }> => {
  // JWTのヘッダーから`kid`を取得
  const decodedJwt = jwt.decode(idToken, { complete: true });
  if (decodedJwt === null || typeof decodedJwt.payload === 'string') {
    // JWTのデコードに失敗した場合は設定不備 or 改竄された可能性があるため500エラー
    debug('verifyIdToken decode error. %s', idToken);
    throw faildDecodeOidcIdToken();
  }
  const kid = decodedJwt.header.kid;

  // JWKSエンドポイントから公開鍵を取得
  const keyStore = await client.issuer.keystore();

  // kidに対応する公開鍵を取得
  const key = keyStore.get({ kid });
  if (!key) {
    // kidに対応する公開鍵がない場合は500エラー
    debug('verifyIdToken mismatch kid. %s', kid);
    throw mismatchKidOidcIdToken();
  }

  // JWTの検証
  try {
    jwt.verify(idToken, key.toPEM(), {
      audience: config.clientId, // OIDCクライアントIDで検証
      issuer: client.issuer.metadata.issuer, // トークン発行者で検証
      algorithms: [key.alg as Algorithm], // アルゴリズムを指定
    });
    // 検証成功の場合はpayloadと期限切れフラグをfalseで返す
    return { payload: decodedJwt.payload as JwtPayload, expired: false };
  } catch (e) {
    debug('verifyIdToken error: %o', e);
    // トークンの有効期限切れの場合のみpayloadと期限切れフラグをtrueで返す
    if (e instanceof TokenExpiredError) {
      debug('Token expired! %s', idToken);
      return { payload: decodedJwt.payload as JwtPayload, expired: true };
    }
    // トークンの有効期限切れ以外は設定不備 or 改竄された可能性があるため500エラー
    throw e;
  }
};

const isRefresh = (expiryDate: number | undefined): boolean => {
  return (
    !expiryDate ||
    Date.now() > (expiryDate - OIDC_TOKEN_REFRESH_BUFFER_SEC) * 1000
  );
};

File: packages/golang/domains/auth/oidc.go
package auth

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"net/http"
	"strings"
	"time"

	"github.com/cam-inc/viron/packages/golang/config"
	"github.com/cam-inc/viron/packages/golang/constant"
	"github.com/cam-inc/viron/packages/golang/domains"
	"github.com/cam-inc/viron/packages/golang/errors"
	"github.com/coreos/go-oidc/v3/oidc"
	"golang.org/x/oauth2"
)

type AuthOIDC struct {
	oidcConfig         *config.OIDC
	oidcProvider       OIDCProvider
	oidcOAuth2Provider OAuthProvider
}

// OIDCProvider インターフェース
type OIDCProvider interface {
	Verifier(config *oidc.Config) *oidc.IDTokenVerifier
	Endpoint() oauth2.Endpoint
}

// OAuthProvider インターフェース
type OAuthProvider interface {
	AuthCodeURL(state string, opts ...oauth2.AuthCodeOption) string
	Exchange(ctx context.Context, code string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error)
	TokenSource(ctx context.Context, token *oauth2.Token) oauth2.TokenSource
}

// OIDCProvider の具体的な実装
type OIDCProviderImpl struct {
	provider *oidc.Provider
}

// Verifier OIDCProviderのIDトークン検証機取得
func (m *OIDCProviderImpl) Verifier(config *oidc.Config) *oidc.IDTokenVerifier {
	return m.provider.Verifier(config)
}

// Endpoint OIDCProviderのエンドポイント取得
func (m *OIDCProviderImpl) Endpoint() oauth2.Endpoint {
	return m.provider.Endpoint()
}

// OAuth2Provider の具体的な実装
type OAuthProviderImpl struct {
	config oauth2.Config
}

// AuthCodeURL OAuth2ProviderのAuthCodeURL
func (m *OAuthProviderImpl) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption) string {
	return m.config.AuthCodeURL(state, opts...)
}

// Exchange OAuth2ProviderのExchange
func (m *OAuthProviderImpl) Exchange(ctx context.Context, code string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
	return m.config.Exchange(ctx, code, opts...)
}

// TokenSource OAuth2ProviderのTokenSource
func (m *OAuthProviderImpl) TokenSource(ctx context.Context, token *oauth2.Token) oauth2.TokenSource {
	return m.config.TokenSource(ctx, token)
}

// newOIDC OIDC認証の初期化
func newOIDC(c config.OIDC, oidcProvider OIDCProvider, oauth2Provider OAuthProvider) *AuthOIDC {
	return &AuthOIDC{
		oidcConfig:         &c,
		oidcProvider:       oidcProvider,
		oidcOAuth2Provider: oauth2Provider,
	}
}

// VerifyAccessToken SSOトークンの検証
func (ao *AuthOIDC) verifyAccessToken(r *http.Request, userID string, ssoToken domains.AdminUserSSOToken) bool {
	log.Info("verifyAccessToken start userId:", userID)
	ctx := r.Context()
	// IDトークンの検証
	// vironではIDトークンはDBに保存されているので、改竄されることはないが
	// DBはviron利用者が管理するため改竄されることを考慮してvironLibとしてはIDトークンの検証を毎度行う
	verifier, err := ao.getTokenVerifier()
	if err != nil {
		log.Errorf("getTokenVerifier failed -> %v", err)
		return false
	}
	idToken, errVerify := verifier.Verify(ctx, ssoToken.IdToken)
	if errVerify == nil {
		// リフレッシュトークンがない場合はIDトークンの有効期限だけで判定
		if ssoToken.RefreshToken == nil && idToken.Expiry.Before(time.Now()) {
			log.Error("IDToken is expired.")
			return false
		}
		// リフレッシュが必要なければtrueを返す
		if !ao.isAccessTokenRefresh(idToken.Expiry) {
			return true
		}
	} else {
		// errVerifyがTokenExpiredErrorの場合でリフレッシュトークンがある場合はリフレッシュ
		if _, ok := errVerify.(*oidc.TokenExpiredError); ok && ssoToken.RefreshToken != nil {
			log.Info("The ID token expired, but I had a refresh token, so I refreshed it.")
		} else {
			log.Errorf("IDToken verify failed -> %v", errVerify)
			return false
		}
	}

	// ---- 以下リフレッシュトークンがある場合の処理 ----
	// リフレッシュトークンがある場合はリフレッシュトークンを使ってトークンを更新
	token := &oauth2.Token{
		AccessToken:  ssoToken.AccessToken,
		TokenType:    ssoToken.TokenType,
		RefreshToken: *ssoToken.RefreshToken,
		Expiry:       time.Unix(0, int64(ssoToken.ExpiryDate)*int64(time.Millisecond)),
	}
	newToken, errRefreshToken := ao.oidcOAuth2Provider.TokenSource(ctx, token).Token()
	if errRefreshToken != nil {
		log.Errorf("TokenSource failed -> %v", err)
		return false
	}

	// IDトークンを取得
	rawIDToken, ok := newToken.Extra("id_token").(string)
	if !ok {
		log.Error("id_token not found")
		return false
	}

	// 新しいクレデンシャルをDBに更新
	ssoToken.AccessToken = newToken.AccessToken
	ssoToken.TokenType = newToken.TokenType
	ssoToken.RefreshToken = &newToken.RefreshToken
	ssoToken.IdToken = rawIDToken
	expiry := int64(newToken.Expiry.UnixNano() / int64(time.Millisecond))
	ssoToken.ExpiryDate = expiry

	// 更新したクレデンシャルをDBに保存
	if err := domains.UpdateAdminUserSSOTokenByUserID(ctx, ao.oidcConfig.ClientID, userID, ssoToken); err != nil {
		log.Errorf("UpdateOneAdminUserSSOTokenByUserID failed -> %v", err)
		return false
	}

	return true
}

// getTokenVerifier IDトークン検証機取得
func (ao *AuthOIDC) getTokenVerifier() (*oidc.IDTokenVerifier, *errors.VironError) {
	oidcConfig := &oidc.Config{
		ClientID: ao.oidcConfig.ClientID,
	}
	return ao.oidcProvider.Verifier(oidcConfig), nil
}

// allowUserHostedDomains emailドメインの許可
func (ao *AuthOIDC) allowUserHostedDomains(email string) bool {
	if len(ao.oidcConfig.UserHostedDomains) == 0 {
		return true
	}
	emailDomain := email[strings.Index(email, "@")+1:]
	for _, v := range ao.oidcConfig.UserHostedDomains {
		if v == emailDomain {
			return true
		}
	}
	return false
}

// isAccessTokenRefresh アクセストークンのリフレッシュ判定
func (ao *AuthOIDC) isAccessTokenRefresh(expiry time.Time) bool {
	// 有効期限の30秒前からリフレッシュ
	return expiry.Add(-constant.OIDC_REFRESH_THRESHOLD).Before(time.Now())
}

// genCodeVerifier Code Verifier生成
func (ao *AuthOIDC) genCodeVerifier() string {
	b := make([]byte, 32)
	_, err := rand.Read(b)
	if err != nil {
		panic(err)
	}
	return base64.RawURLEncoding.EncodeToString(b)
}

// genAuthorizationUrl Authorization URL生成
func (ao *AuthOIDC) genAuthorizationUrl(codeVerifier string, state string) string {
	options := []oauth2.AuthCodeOption{
		oauth2.S256ChallengeOption(codeVerifier),
	}

	// Googleの場合は特殊なパラメータを追加
	if ao.isGoogle() {
		options = append(options, oauth2.AccessTypeOffline, oauth2.ApprovalForce)
	}

	return ao.oidcOAuth2Provider.AuthCodeURL(state, options...)
}

// signin OIDC認証
func (ao *AuthOIDC) signin(r *http.Request, code string, state string, codeVerifier string, multipleAuthUser bool) (string, *errors.VironError) {
	ctx := r.Context()

	// 許可コードとトークンを交換
	oidcToken, errExchange := ao.oidcOAuth2Provider.Exchange(
		ctx,
		code,
		oauth2.SetAuthURLParam("code_verifier", codeVerifier),
		oauth2.SetAuthURLParam("state", state),
	)
	if errExchange != nil {
		log.Errorf("Exchange failed -> %v", errExchange)
		return "", errors.SigninFailed
	}

	// IDトークンを取得
	rawIDToken, ok := oidcToken.Extra("id_token").(string)
	if !ok {
		log.Error("id_token not found")
		return "", errors.SigninFailed
	}

	// IDトークン検証機を取得
	verifier, errVerifier := ao.getTokenVerifier()
	if errVerifier != nil {
		log.Errorf("getTokenVerifier failed -> %v", errVerifier)
		return "", errors.SigninFailed
	}

	// IDトークンを検証
	idToken, errVerify := verifier.Verify(ctx, rawIDToken)
	if errVerify != nil {
		log.Errorf("Verify failed -> %v", errVerify)
		return "", errors.SigninFailed
	}

	// email取得
	var claims struct {
		Email string `json:"email"`
	}
	if err := idToken.Claims(&claims); err != nil {
		log.Errorf("Claims failed -> %v", err)
		return "", errors.SigninFailed
	}

	// emailが取得できない場合はエラー
	if claims.Email == "" {
		log.Error("Email not found")
		return "", errors.SigninFailed
	}

	// emailドメインが許可されているか確認
	if !ao.allowUserHostedDomains(claims.Email) {
		log.Error("domainCheck is false.")
		return "", errors.SigninFailed
	}

	// ユーザーが存在しない場合は作成
	user := domains.FindByEmail(ctx, claims.Email)

	// SSOトークンはUpsert
	provider := constant.AUTH_SSO_IDPROVIDER_CUSTOME
	if ao.isGoogle() {
		provider = constant.AUTH_SSO_IDPROVIDER_GOOGLE
	}
	expiry := int64(oidcToken.Expiry.UnixNano() / int64(time.Millisecond))
	ssoToken := &domains.AdminUserSSOToken{
		AuthType:     constant.AUTH_TYPE_OIDC,
		Provider:     provider,
		ClientID:     ao.oidcConfig.ClientID,
		TokenType:    oidcToken.TokenType,
		IdToken:      rawIDToken,
		AccessToken:  oidcToken.AccessToken,
		RefreshToken: &oidcToken.RefreshToken,
		ExpiryDate:   expiry,
	}

	if user == nil {
		// userが存在しない場合
		// adminUser Entity
		userPayload := &domains.AdminUser{
			Email: claims.Email,
		}
		var err error
		// 初回の管理者ユーザー作成
		user, ssoToken, err = createFirstAdminUser(ctx, userPayload, ssoToken, constant.AUTH_TYPE_OIDC)
		if err != nil {
			log.Errorf("create first admin user failed err:%v, user:%v", err, user)
			return "", errors.SigninFailed
		}

		// 管理者ユーザー存在する場合はviewerユーザー作成
		if user == nil {
			if user, ssoToken, err = createViewer(ctx, userPayload, ssoToken, constant.AUTH_TYPE_OIDC); err != nil {
				log.Errorf("create admin user(viewer) failed err:%v, user:%v ssoToken", err, user, ssoToken)
			}
		}
	} else {
		// userが存在する場合
		// multipleAuthUser=falseの場合は認証方法は1つのみに制限する
		if !multipleAuthUser {
			// 登録済みユーザーの認証方法の確認
			// password認証が登録済みの場合はエラー
			if user.Password != nil {
				log.Error("user already exists with password.")
				return "", errors.SigninFailed
			}
			// userIdで登録済みSSOトークン情報を取得
			ssoTokens, err := domains.ListAdminUserSSOToken(ctx, &domains.AdminUserSSOTokenConditions{UserID: user.ID, Size: 99})
			if err != nil {
				log.Errorf("ListAdminUserSSOToken failed err:%v", err)
				return "", errors.SigninFailed
			}
			// 今回のclientIDと違うSSOトークンがある場合はエラー
			for _, v := range ssoTokens.List {
				if v.ClientID != ssoToken.ClientID {
					log.Error("user already exists with oidc.")
					return "", errors.SigninFailed
				}
			}
		}

		// SSOトークンのUserIDを設定
		ssoToken.UserID = user.ID

		// ここまででエラーがない場合はSSOトークンのUpsert
		if err := domains.UpsertAdminUserSSOToken(ctx, ssoToken); err != nil {
			log.Errorf("UpsertAdminUserSSOToken failed err:%v", err)
			return "", errors.SigninFailed
		}
	}

	// JWTトークンを作成
	token, errSign := Sign(r, user.ID)
	if errSign != nil {
		log.Error("Signin sign failed %#v \n", errSign)
		return "", errors.SigninFailed
	}
	return token, nil
}

// isGoogle GoogleのOIDCかどうか
func (ao *AuthOIDC) isGoogle() bool {
	return ao.oidcConfig.IssuerURL == constant.GOOGLE_OIDC_ISSUER_URL
}
